cajita/src/Cajita_ParticleGridDistributor.hpp

```diff
@@ -0,0 +1,390 @@
+    // this list may be invalid.
+    template <std::size_t NSD = num_space_dim>
+    std::enable_if_t<2 == NSD, std::vector<int>> getTopology()
+    {
+        std::vector<int> topology( 9, -1 );
+        int nr = 0;
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i, ++nr )
+                topology[nr] = _local_grid.neighborRank( i, j );
+        return topology;
+    }
+
+    // Locate the particles in the local grid and get their destination rank.
+    // Particles are assumed to only migrate to a location in the nearest
+    // neighbor halo or stay on this rank. If the particle crosses a global
+    // periodic boundary, wrap it's coordinates back into the domain.
+    template <class PositionSliceType, class NeighborRankView,
+              class DestinationRankView>
+    void getMigrateDestinations( const NeighborRankView& neighbor_ranks,
+                                 DestinationRankView& destinations,
+                                 PositionSliceType& positions )
+    {
+        using execution_space = typename PositionSliceType::execution_space;
+
+        // Check within the local domain.
+        const auto& local_mesh =
+            Cajita::createLocalMesh<Kokkos::HostSpace>( _local_grid );
+
+        // Use global domain for periodicity.
+        const auto& global_grid = _local_grid.globalGrid();
+        const auto& global_mesh = global_grid.globalMesh();
+
+        Kokkos::Array<double, num_space_dim> local_low{};
+        Kokkos::Array<double, num_space_dim> local_high{};
+        Kokkos::Array<bool, num_space_dim> periodic{};
+        Kokkos::Array<double, num_space_dim> global_low{};
+        Kokkos::Array<double, num_space_dim> global_high{};
+        Kokkos::Array<double, num_space_dim> global_extent{};
+
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            local_low[d] = local_mesh.lowCorner( Cajita::Own(), d );
+            local_high[d] = local_mesh.highCorner( Cajita::Own(), d );
+            periodic[d] = global_grid.isPeriodic( d );
+            global_low[d] = global_mesh.lowCorner( d );
+            global_high[d] = global_mesh.highCorner( d );
+            global_extent[d] = global_mesh.extent( d );
+        }
+
+        Kokkos::parallel_for(
+            "get_migrate_destinations",
+            Kokkos::RangePolicy<execution_space>( 0, positions.size() ),
+            KOKKOS_LAMBDA( const int p ) {
+                // Compute the logical index of the neighbor we are sending to.
+                int nid[num_space_dim];
+                for ( std::size_t d = 0; d < num_space_dim; ++d )
+                {
+                    nid[d] = 1;
+                    if ( positions( p, d ) < local_low[d] )
+                        nid[d] = 0;
+                    else if ( positions( p, d ) > local_high[d] )
+                        nid[d] = 2;
+                }
+
+                // Compute the destination MPI rank [ni + 3*(nj + 3*nk) in 3d].
```

Comment:
this should also work in 2d right?

Reply:
Yes, I just left the comment because the code isn't as readable for 2d & 3d